src/server/qgsconfigcache.cpp

```diff
@@ -48,9 +49,29 @@ const QgsProject *QgsConfigCache::project( const QString &path )
     {
       if ( !badLayerHandler->badLayers().isEmpty() )
       {
-        QString errorMsg = QStringLiteral( "Layer(s) %1 not valid" ).arg( badLayerHandler->badLayers().join( ',' ) );
-        QgsMessageLog::logMessage( errorMsg, QStringLiteral( "Server" ), Qgis::Critical );
-        throw QgsServerException( QStringLiteral( "Layer(s) not valid" ) );
+        // if bad layers are not restricted layers so service failed
+        QStringList unrestrictedBadLayers;
+        // test bad layers through restrictedlayers
+        const QStringList badLayerIds = badLayerHandler->badLayers();
+        const QMap<QString, QString> badLayerNames = badLayerHandler->badLayerNames();
+        const QStringList resctrictedLayers = QgsServerProjectUtils::wmsRestrictedLayers( *prj );
+        for ( const QString &badLayerId : badLayerIds )
+        {
+          // if this bad layer is in restricted layers
+          // it doesn't need to be added to unrestricted bad layers
+          if ( badLayerNames.contains( badLayerId ) &&
+               resctrictedLayers.contains( badLayerNames.value( badLayerId ) ) )
+          {
+            continue;
+          }
+          unrestrictedBadLayers.append( badLayerId );
+        }
+        if ( !unrestrictedBadLayers.isEmpty() )
```

Comment:
For the sake of performance, you don't need to loop over all layers, store all unrestricted bad layers, *then* throw if unrestrictedBadLayers is not empty: all you have to do is   just throw at the first unrestricted bad layer.

Reply:
yes but for the message I need to know all unrestricted bad layers.